### PR TITLE
[One .NET] Profiled AOT support

### PR DESCRIPTION
**Pull request by jonathanpeppers**
Fixes: https://github.com/xamarin/xamarin-android/issues/6053

This enables the ability to set:

    <PropertyGroup Condition=" '$(Configuration)' == 'Release' ">
      <RunAOTCompilation>true</RunAOTCompilation>
      <AndroidEnableProfiledAot>true</AndroidEnableProfiledAot>
    </PropertyGroup>

And a default `dotnet.aotprofile` will be passed to the
`<MonoAOTCompiler/>` MSBuild task.

There will not be a way to "record" profiles in the Android workload
in .NET 6. This is because it relies on legacy Mono infrastructure
that is not implemented in .NET 6.

I created a Github repo for recording profiles, where I committed the
necessary binaries such as `aprofutil` and `libmono-profiler-aot.so`:

https://github.com/jonathanpeppers/android-profiled-aot

I recorded a `dotnet.aotprofile` with the contents:

    Modules:
        6799590C-2963-4835-AEDC-20C67D875132 System.Private.CoreLib
        2CC403AF-60DC-48A9-9B69-FF254277AEA0 Mono.Android
        92FF7068-2EA9-4681-B340-44E91077417A Java.Interop
        008C0F24-5014-4D41-AAB9-DDFCEA59E171 Xamarin.Google.Android.Material
        7CEE6098-BE9F-4043-B35A-B8B0684EF0CB Xamarin.AndroidX.AppCompat
        20D1FBB8-829A-4618-8E82-8047C8827EA8 Xamarin.AndroidX.Fragment
        CC363D7A-9CDB-4999-9E10-279412B14A1B Xamarin.AndroidX.Activity
        21CF52B7-0256-4838-848C-DA026B2F1789 Xamarin.AndroidX.Core
        D1DE5607-BC27-45FC-93EC-0542C787E5FF Xamarin.AndroidX.DrawerLayout
    Summary:
        Modules:          9
        Types:          229
        Methods:      1,010

`AndroidApp1` is the `Navigation Drawer App` template from "legacy"
Xamarin.Android. I ported this template to .NET 6 and dropped usage of
Xamarin.Essentials. I thought this was a good target for a default
profile, because of its heavy usage of AndroidX and Google Material.

In a future PR, I will add a default AOT profile for .NET MAUI to be
shipped inside the `maui` workload.

## Results ##

All tests:

 1. Were running on a [Google Pixel 5][0], and
 2. Enabled two architectures, arm64 and x86, and
 3. **JIT time** was average of 10 runs with `-c Release`, no AOT
 4. **AOT time** was average of 10 runs with `-c Release
    -p:RunAOTCompilation=true`, with the`Activity: Displayed` time
 5. **Profiled AOT time** was average of 10 runs with `-c Release
    -p:RunAOTCompilation=true -p:AndroidEnableProfiledAot=true` with
    the `Activity: Displayed` time.

| Test               |    JIT time |    AOT time | Profiled AOT time | JIT apk size  |  AOT apk size | Profiled AOT apk size |
| ------------------ | ----------: | ----------: | ----------------: | ------------: | ------------: | --------------------: |
| [AndroidApp1][1]   |  00:00.4387 |  00:00:3317 |        00:00.3093 |     9,155,954 |    12,755,672 |             9,777,880 |
| [MauiApp1][2]      |  00:01.4205 |  00:00:7285 |        00:00.7098 |    17,435,225 |    44,751,651 |            23,210,787 |

[0]: store.google.com/us/product/pixel_5_specs?hl=en-US
[1]: https://github.com/jonathanpeppers/android-profiled-aot/tree/main/AndroidApp1
[2]: https://github.com/jonathanpeppers/android-profiled-aot/tree/main/MauiApp1

### CODE REVIEW

**Comment by charlesroddie** [comment]
> HelloMaui ... 43,442,233

The moves from mono to monovm+corefx, from Xamarin to Maui, and the work in dotnet on linkers all have had no effect?

**Comment by jonathanpeppers** [comment]
@charlesroddie the HelloMaui full AOT app size is bigger (from what we've seen), due to two things:

* The AOT compiler's output is larger (iOS team had similar results)
* dotnet/maui is not setup to be Trimmable yet

We hope these numbers will get better by the time .NET 6 is released.

File: src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets [hunk 1]
```diff
@@ -680,6 +680,12 @@ because xbuild doesn't support framework reference assemblies.
 		/>
 	</CreateProperty>
 
+	<CreateProperty Value="$(MonoAndroidBinDirectory)">
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets:683]: Should we start using `<PropertyGroup/>` here? :-)

File: src/monodroid/jni/monodroid-glue.cc [hunk 2]
```diff
@@ -1582,23 +1581,10 @@ MonodroidRuntime::set_profile_options ()
...
-	// path without allocation
-	dynamic_local_string<SENSIBLE_PATH_MAX + PROPERTY_VALUE_BUFFER_LEN> value;
-	{
-		dynamic_local_string<PROPERTY_VALUE_BUFFER_LEN> prop_value;
-		if (androidSystem.monodroid_get_system_property (Debug::DEBUG_MONO_PROFILE_PROPERTY, prop_value) == 0)
-			return;
-
-		value.assign (prop_value.get (), prop_value.length ());
-	}
+#endif // def NET6
```
Comment (grendello) [inline src/monodroid/jni/monodroid-glue.cc:1587]: Could we instead copy the code that deals with the `aotprofile` option to the net6 version of `set_profile_options`? None of the other options work in net6+. Copying the code will make it clear what is and what isn't supported in net6 and will also make it easier to just delete the "legacy" option once we drop the support.

File: src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets (hunk 1 above)
Comment (jonathanpeppers) [reply src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets]: I started a new `<PropertyGroup/>` and we can port all these in the future.

File: src/monodroid/jni/monodroid-glue.cc (hunk 2 above)
Comment (jonathanpeppers) [reply src/monodroid/jni/monodroid-glue.cc]: I think I did it right? Let me know if there is code that can be removed.

**Comment by jonpryor** [comment]
```markdown
Fixes: https://github.com/xamarin/xamarin-android/issues/6053

Context: https://github.com/jonathanpeppers/android-profiled-aot

This enables the ability to set:

	<PropertyGroup Condition=" '$(Configuration)' == 'Release' ">
	  <RunAOTCompilation>true</RunAOTCompilation>
	  <AndroidEnableProfiledAot>true</AndroidEnableProfiledAot>
	</PropertyGroup>

Add a default `dotnet.aotprofile` which will be passed to the
`<MonoAOTCompiler/>` MSBuild task.

There will not be a way to "record" profiles in the Android workload
in .NET 6. This is because it relies on legacy Mono infrastructure
that is not implemented in .NET 6.

I created a Github repo for recording profiles, where I committed the
necessary binaries such as `aprofutil` and `libmono-profiler-aot.so`:

  * https://github.com/jonathanpeppers/android-profiled-aot

I recorded a `dotnet.aotprofile` with the contents:

	Modules:
	    6799590C-2963-4835-AEDC-20C67D875132 System.Private.CoreLib
	    2CC403AF-60DC-48A9-9B69-FF254277AEA0 Mono.Android
	    92FF7068-2EA9-4681-B340-44E91077417A Java.Interop
	    008C0F24-5014-4D41-AAB9-DDFCEA59E171 Xamarin.Google.Android.Material
	    7CEE6098-BE9F-4043-B35A-B8B0684EF0CB Xamarin.AndroidX.AppCompat
	    20D1FBB8-829A-4618-8E82-8047C8827EA8 Xamarin.AndroidX.Fragment
	    CC363D7A-9CDB-4999-9E10-279412B14A1B Xamarin.AndroidX.Activity
	    21CF52B7-0256-4838-848C-DA026B2F1789 Xamarin.AndroidX.Core
	    D1DE5607-BC27-45FC-93EC-0542C787E5FF Xamarin.AndroidX.DrawerLayout
	Summary:
	    Modules:          9
	    Types:          229
	    Methods:      1,010

`AndroidApp1` is the `Navigation Drawer App` template from "legacy"
Xamarin.Android.  I ported this template to .NET 6 and dropped usage
of Xamarin.Essentials. I thought this was a good target for a default
profile, because of its heavy usage of AndroidX and Google Material.

In a future PR, I will add a default AOT profile for .NET MAUI to be
shipped inside the `maui` workload.

~~ Results ~~

All tests:

 1. Were running on a [Google Pixel 5][0], and
 2. Enabled two architectures, arm64 and x86, and
 3. **JIT time** was average of 10 runs with `-c Release`, no AOT
 4. **AOT time** was average of 10 runs with `-c Release
    -p:RunAOTCompilation=true`, with the`Activity: Displayed` time
 5. **Profiled AOT time** was average of 10 runs with `-c Release
    -p:RunAOTCompilation=true -p:AndroidEnableProfiledAot=true` with
    the `Activity: Displayed` time.

|                                     |    [AndroidApp1][1] |     [MauiApp1][2] |
| ----------------------------------: | ------------------: | ----------------: |
|                JIT startup time (s) |   00:00.4387        | 00:01.4205        |
|          AOT startup time (vs. JIT) |   00:00.3317 ( 76%) | 00:00.7285 ( 51%) |
| Profiled AOT startup time (vs. JIT) |   00:00.3093 ( 71%) | 00:00.7098 ( 50%) |
|                 JIT `.apk` size (B) |    9,155,954        | 17,435,225        |
|           AOT `.apk` size (vs. JIT) |   12,755,672 (139%) | 44,751,651 (257%) |
|  Profiled AOT `.apk` size (vs. JIT) |    9,777,880 (107%) | 23,210,787 (133%) |

[0]: store.google.com/us/product/pixel_5_specs?hl=en-US
[1]: jonathanpeppers/android-profiled-aot@e48c6df/AndroidApp1
[2]: jonathanpeppers/android-profiled-aot@e48c6df/MauiApp1

Co-authored-by: Marek Habersack <grendel@twistedcode.net>
```

**Pull request merged**